### PR TITLE
feat!: introduce Gramian row-mean

### LINKED ISSUE #599: Refactor `kernel_matrix_row_sum` methods
### What's the new feature?

Rename `calculate_kernel_matrix_row_sum_mean` to `gramian_row_mean` and utilise `jax.lax.scan` to enhance compile time performance. Additionally remove `update_kernel_matrix_row_sum` and `calculate_kernel_matrix_row_sum` methods as these are only used by `calculate_kernel_matrix_row_sum_mean` (they don't need to be a public part of the API).

### What value does this add?

- A more informative and concise name for the `kernel_matrix_row_sum_mean`
- Much faster compilation times
- Reduction in the size of the public API. Reduces the amount of code that must be tested/maintained.

### Is there an alternative you've considered?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tc85324**
### PR Type
Closes #599

- Bugfix
- Feature
- Code style update (formatting, local variables)
- Refactoring (no functional changes)
- Documentation content changes
- Tests

Renames the `kernel_matrix_row_sum_mean` to the more concise Gramian row-mean, and utilises `jax.lax.scan` to provide more control over the block iterated computation of this quantity.

More comprehensive testing is now performed when checking the behaviour of various `block_size` (previously called `max_size`) values, when passed to `gramian_row_mean`. As a result, some other tests have become redundant, and have been removed (primarily in `test_coresubset.py`).

### Description

(Write your answer here.)

### How Has This Been Tested?
All tests pass

### Does this PR introduce a breaking change?
removed `{update,calculate}_kernel_matrix_row_sum` and `calculate_kernel_matrix_row_sum_mean`; replaced by calls to `gramian_row_mean`. All references to `kernel_matrix_row_sum_mean` have been replaced by `gramian_row_mean`. The `KernelComputeType` has also been removed as it is now only used in `coresubset.py`.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: examples/herding_approximate_gram_matrix.py [hunk 1]
```diff
@@ -19,10 +19,9 @@
 points sampled from ``k`` clusters in space.
 
 A coreset is generated using kernel herding, with a Squared Exponential kernel. To
-reduce computational demand, we approximate the kernel matrix row sum mean (also known
-as the mean of the Gram matrix row sum). This coreset is compared to a coreset generated
-via uniform random sampling. Coreset quality is measured using maximum mean discrepancy
-(MMD).
+reduce computational demand, we approximate the kernel's Gramian row mean. This coreset
```
Comment (db091756) [inline examples/herding_approximate_gram_matrix.py:22]: "kernel's Gramian row mean" -> "kernel Gramian's row mean"?

File: examples/herding_approximate_gram_matrix.py [hunk 2]
```diff
@@ -54,13 +53,13 @@
 # pylint: disable=too-many-locals
 def main(out_path: Path | None = None) -> tuple[float, float]:
     """
-    Run the tabular data herding example with an approximate kernel matrix row sum mean.
+    Run the tabular data herding example with an approximate kernel Gramian row-mean.
```
Comment (db091756) [inline examples/herding_approximate_gram_matrix.py:56]: You don't use the hyphen elsewhere so probably remove to keep consistent, same for 60

File: examples/herding_approximate_gram_matrix.py (hunk 1 above)
Comment (db091756) [inline examples/herding_approximate_gram_matrix.py:22]: row mean needs hyphen?

File: coreax/approximation.py [hunk 3]
```diff
@@ -169,19 +169,18 @@ class MonteCarloApproximateKernel(RandomRegressionKernel):
...
-        Calculate the kernel matrix row sum mean by Monte Carlo approximation.
+        Approximate the Gramian row-mean by Monte-Carlo sampling.
 
-        A uniform random subset of ``x`` is used to approximate the kernel matrix row
-        sum mean.
+        A uniform random subset of ``x`` is used to approximate the base kernel's
+        Gramian row-mean.
 
         :param x: Data matrix, :math:`n \times d`
         :param max_size: Size of matrix block to process
```
Comment (db091756) [inline coreax/approximation.py:180]: Should be block_size?

File: coreax/approximation.py [hunk 4]
```diff
@@ -210,20 +209,19 @@ class ANNchorApproximateKernel(RandomRegressionKernel):
...
-        Calculate the kernel matrix row sum mean by random regression on ANNchor points.
+        Approximate the Gramian row-mean by random regression on ANNchor points.
 
         A subset of ``x`` is selected via the ANNchor approach and random kernel
-        regression used to approximate the kernel matrix row sum mean. The ANNchor
+        regression used to approximate the base kernel's Gramian row-mean. The ANNchor
         implementation used can be found `here <https://github.com/gchq/annchor>`_.
 
         :param x: Data matrix, :math:`n \times d`
         :param max_size: Size of matrix block to process
```
Comment (db091756) [inline coreax/approximation.py:221]: Should be block_size?

File: coreax/approximation.py [hunk 5]
```diff
@@ -257,32 +255,30 @@ class NystromApproximateKernel(RandomRegressionKernel):
...
         We consider a :math:`n \times d` dataset, and wish to use an :math:`m \times d`
-        subset of this to approximate the kernel matrix row sum mean. The ``m`` points
-        are selected uniformly at random, and the Nystrom estimator, as defined in
-        :cite:`chatalic2022nystrom` is computed using this subset.
+        subset of this to approximate the base kernel's Gramian row-mean. The ``m``
+        points are selected uniformly at random, and the Nystrom estimator, as defined
+        in :cite:`chatalic2022nystrom` is computed using this subset.
 
         :param x: Data matrix, :math:`n \times d`
         :param max_size: Size of matrix block to process
```
Comment (db091756) [inline coreax/approximation.py:278]: Should be block_size?

File: tests/unit/test_refine.py [hunk 6]
```diff
@@ -159,7 +159,7 @@ def test_refine(
 
         When ``use_cached_row_sum_mean=True``, we expect the corresponding cached value
         in the coreset object to be used by refine, otherwise, we expect the kernel's
-        calculate_kernel_matrix_row_sum_mean to be called (exactly once).
```
Comment (db091756) [inline tests/unit/test_refine.py:162]: Maybe refer to use_cached_row_mean rather than use_cached_row_sum_mean now?

**Review by db091756 — CHANGES_REQUESTED** [review]
Looks good to me, just a few missed docs changes I think, Ctrl+Shift+f "max_size" and "row_sum_mean" flags a few issues, highlighted some of them. Otherwise all good

File: examples/herding_approximate_gram_matrix.py (hunk 1 above)
Comment (tc85324) [reply examples/herding_approximate_gram_matrix.py:22]: Yes it does, nice spot!

File: coreax/approximation.py (hunk 3 above)
Comment (tc85324) [reply coreax/approximation.py:180]: Yes, I'll double check they're aren't any more of these! Thanks for spotting!

File: tests/unit/test_refine.py (hunk 6 above)
Comment (tc85324) [reply tests/unit/test_refine.py:162]: Agreed

**Review by db091756 — APPROVED** [review]

**Pull request merged**